keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,965 @@
+                if !pluginStatus.IsSuccess() {
+                        if !pluginStatus.IsUnschedulable() {
+```
+
+##### Resulting output
+
+Here is log output from `kube-scheduler -v5` for a Pod with an inline volume which cannot be created because storage is exhausted:
+
+```
+I1026 16:21:00.461394  801139 scheduler.go:436] "Attempting to schedule pod" pod="default/my-csi-app-inline-volume"
+I1026 16:21:00.461476  801139 binder.go:730] PreFilter/VolumeBinding: "PVC is not bound" pod="default/my-csi-app-inline-volume" pvc="default/my-csi-app-inline-volume-my-csi-volume"
+```
+
+The next line is from a file which has not been converted. It’s not clear in which context that message gets emitted:
+````
+I1026 16:21:00.461619  801139 csi.go:222] "Persistent volume had no name for claim" PVC="default/my-csi-app-inline-volume-my-csi-volume"
+````
+
+```
+I1026 16:21:00.461647  801139 binder.go:266] NominatedPods/Filter/VolumeBinding: "FindPodVolumes starts" pod="default/my-csi-app-inline-volume" node="127.0.0.1"
+I1026 16:21:00.461673  801139 binder.go:842] NominatedPods/Filter/VolumeBinding: "No matching volumes for PVC on node" pod="default/my-csi-app-inline-volume" node="127.0.0.1" default/my-csi-app-inline-volume-my-csi-volume="127.0.0.1"
+I1026 16:21:00.461724  801139 binder.go:971] NominatedPods/Filter/VolumeBinding: "Node has no accessible CSIStorageCapacity with enough capacity" pod="default/my-csi-app-inline-volume" node="127.0.0.1" pvc="default/my-csi-app-inline-volume-my-csi-volume" pvcSize=549755813888000 sc="csi-hostpath-fast"
+I1026 16:21:00.461817  801139 preemption.go:195] "Preemption will not help schedule pod on any node" pod="default/my-csi-app-inline-volume"
+I1026 16:21:00.461886  801139 scheduler.go:464] "Status after running PostFilter plugins for pod" pod="default/my-csi-app-inline-volume" status=&{code:2 reasons:[0/1 nodes are available: 1 Preemption is not helpful for scheduling.] err:<nil> failedPlugin:}
+I1026 16:21:00.461918  801139 factory.go:209] "Unable to schedule pod; no fit; waiting" pod="default/my-csi-app-inline-volume" err="0/1 nodes are available: 1 node(s) did not have enough free storage."
+```
+
+### Test Plan
+
+The new code will be covered by unit tests that execute as part of `pull-kubernetes-unit`.
+
+Converted components will be tested by exercising them with JSON output at high
+log levels to emit as many log messages as possible. Analysis of those logs
+will detect duplicate keys that might occur when a caller uses `WithValues` and
+a callee adds the same value in a log message.
+
+Static code analysis cannot detect this, or at least not easily. What it can
+check is that the individual log calls pass valid key/value pairs.
+
+
+### Graduation Criteria
+
+#### Alpha
+
+- Common utility code available
+- At least kube-scheduler framework and some plugins converted
+- Initial e2e tests completed and enabled in Prow
+
+#### Beta
+
+- All of kube-scheduler converted
```

Comment:
If we also start to convert at least one thing that typically runs outside the control plane, we'll get better feedback that could help inform the remaining work. This formal choice to pick one extra component outside the scheduler could be for either alpha or beta.

Reply:
What could that other thing be? There's not much in k/k that isn't part of the control plane.

kubectl might be a candidate, but that typically doesn't log much at all (basically just warnings from client-go), so the effect of converting it wouldn't be very visible.

Perhaps we need to consider components outside of k/k? I'm open for suggestions.

Reply:
> components outside of k/k

My (uninformed) suggestion is the cluster autoscaler. It's something people run both in-cluster and separately, and it ought to be able to write logs with context. I don't know what in-project, out-of-tree components have already made headway in adopting structured logging.

Reply:
That's a good suggestion. It is based largely on the scheduler code and thus will automatically inherit all changes that we make there.

Reply:
I think kube-proxy falls into the same bucket as "Kubernetes control plane" - users who check the control plane are also likely to check kube-proxy and vice versa.

My understanding of @sftim's proposal was to branch out and reach people who normally do not pay attention to the control plane pods. Perhaps the autoscaler doesn't go far away enough from that, but I think it goes further than kube-proxy.

I know from SIG Scaling that users have problems configuring and using the autoscaler. It might be that improved logs could help with that problem.

Reply:
I guess it's challenging in AWS. On GKE it's managed, so we have the same "control-plane" problem.

Reply:
Another alternative would be the CSI sidecars. I already did a [prototype for external-provisioner](https://github.com/kubernetes-csi/external-provisioner/pull/639) without using any of the APIs proposed here. That one could be adapted. The advantage is that there is interest in structured logging, we (= SIG-Storage) just decided to wait until there is also a decision about contextual logging.

Reply:
Let me go with the external-provisioner as another component for beta.